pkg/runtime/gomemlimit_linux.go

```diff
@@ -3,49 +3,141 @@
 // This product includes software developed at Datadog (https://www.datadoghq.com/).
 // Copyright 2016-present Datadog, Inc.
 
-//go:build linux && go1.19
+//go:build linux
 
 package runtime
 
 import (
+	"context"
 	"errors"
-	"os"
+	"math"
 	"runtime/debug"
+	"time"
 
 	"github.com/DataDog/datadog-agent/pkg/util/cgroups"
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
-// SetGoMemLimit configures Go memory soft limit based on cgroups.
-// The soft limit is set to 90% of the cgroup memory hard limit.
-// The function is noop if
-//   - GOMEMLIMIT is set already
-//   - There is no cgroup limit
-//
-// Read more about Go memory limit in https://tip.golang.org/doc/gc-guide#Memory_limit
-func SetGoMemLimit(isContainerized bool) error {
-	if _, ok := os.LookupEnv("GOMEMLIMIT"); ok {
+type staticMemoryLimiter struct {
+	limitPct        float64
+	isContainerized bool
+}
+
+// NewStaticMemoryLimiter return a memory limit that sets memory limit once based on cgroup limit
+func NewStaticMemoryLimiter(limitPct float64, isContainerized bool) MemoryLimiter {
+	return staticMemoryLimiter{
+		limitPct:        limitPct,
+		isContainerized: isContainerized,
+	}
+}
+
+// Run runs the memory limit logic
+func (s staticMemoryLimiter) Run(context.Context) error {
+	// SetMemoryLimit with negative values returns current value
+	if debug.SetMemoryLimit(-1) != math.MaxInt64 {
 		log.Debug("GOMEMLIMIT is set already, doing nothing")
 		return nil
 	}
-	selfReader, err := cgroups.NewSelfReader("/proc", isContainerized)
+
+	selfCgroup, err := getSelfCgroup(s.isContainerized)
 	if err != nil {
 		return err
 	}
-	cgroup := selfReader.GetCgroup(cgroups.SelfCgroupIdentifier)
-	if cgroup == nil {
-		return errors.New("cannot get cgroup")
-	}
+
 	var stats cgroups.MemoryStats
-	if err := cgroup.GetMemoryStats(&stats); err != nil {
+	if err := selfCgroup.GetMemoryStats(&stats); err != nil {
 		return err
 	}
 	if stats.Limit == nil {
-		log.Debug("Cgroup memory limit not found, doing nothing")
 		return nil
 	}
-	softLimit := int64(0.9 * float64(*stats.Limit))
-	log.Infof("Cgroup memory limit is %d, setting gomemlimit to %d", *stats.Limit, softLimit)
-	debug.SetMemoryLimit(softLimit)
+
+	s.setMemoryLimit(*stats.Limit)
 	return nil
 }
+
+func (s staticMemoryLimiter) setMemoryLimit(cgroupLimit uint64) {
+	memLimit := int64(s.limitPct * float64(cgroupLimit))
+	log.Infof("Cgroup memory limit is %d, setting gomemlimit to %d", cgroupLimit, memLimit)
+	debug.SetMemoryLimit(memLimit)
+}
+
+type dynamicMemoryLimiter struct {
+	ticker               *time.Ticker
+	selfCgroup           cgroups.Cgroup
+	minLimitPct          float64
+	externalMemoryReader func(cgroups.MemoryStats) uint64
+}
+
+// NewDynamicMemoryLimiter creates a memory limiter that adapts limit continuously based on Cgroup limit
+// plus memory consumed outside of the Go space.
+// It will set GOMEMLIMT to max(cgroupLimit * minLimitPct, cgroupLimit-externalMemory)
+// It will override existing value of GOMEMLIMIT
```

Comment:
Shouldn't it be the contrary? As in something provided in the environment var with GOMEMLIMIT should override the shipped logic?